### PR TITLE
Feature/popover

### PR DESCRIPTION
**Pull request by Tibing**
### Please read and mark the following check list before creating a pull request:

 - [x] I read and followed the [CONTRIBUTING.md](https://github.com/akveo/nebular/blob/master/CONTRIBUTING.md) guide.
 - [x] I read and followed the [New Feature Checklist](https://github.com/akveo/nebular/blob/master/DEV_DOCS.md#new-feature-checklist) guide.
 
 #### Short description of what this resolves:

Add popover directive

### CODE REVIEW

File: src/framework/theme/components/popover/popover.component.scss [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  $arrow-size: 11px;
+  $arrow-offset: -($arrow-size * 2);
+
+  position: absolute;
+  z-index: 1000;
+  border-radius: 5px;
+  top: 200px;
+
+  .primitive-popover {
+    padding: 0.375rem 3rem;
```
Comment (nnixaa) [inline src/framework/theme/components/popover/popover.component.scss]: padding: 0.75rem 1rem;, approved by Nick

File: src/framework/theme/components/popover/popover.directive.ts [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+ * @example Custom components
+ *
+ * ```
+ * <button [nbPopover]="NbCardComponent"></button>
+ * ```
+ *
+ * @example Primitive types
+ *
+ * ```
+ * <button [nbPopover]="'Hello, Popover!'"></button>
```
Comment (nnixaa) [inline src/framework/theme/components/popover/popover.directive.ts]: `<button nbPopover="Hello, Popover!"></button>`
no need to use [] and additional '' in all of your examples

File: src/framework/theme/components/popover/popover.directive.ts [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+
+/**
+ * Popover can be one of the following types:
+ * template, component or plain js string.
+ * So NbPopoverContent provides types alias for this purposes.
+ * */
+export type NbPopoverContent = string | TemplateRef<any> | Type<any>;
+
+/**
+ * Powerful popover directive, which provides the best UX for your users.
```
Comment (nnixaa) [inline src/framework/theme/components/popover/popover.directive.ts]: Why don't we add a beautiful gif as on the Flip Card page

File: src/app/app.module.ts [hunk 4]
```diff
@@ -73,6 +74,7 @@ import { NbFormsTestComponent } from './forms-test/forms-test.component';
 import { NbCardTestComponent } from './card-test/card-test.component';
 import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
 import { AuthGuard } from './auth-guard.service';
+import { NbPopoverTestComponent } from './popover-test/popover-test.component';
```
[the remark this answers is not in the document]
Comment (Tibing) [reply src/app/app.module.ts]: fixed

File: src/app/app.routes.ts [hunk 5]
```diff
@@ -50,6 +50,7 @@ import {
   NbResetPasswordComponent,
 } from '@nebular/auth';
 import { AuthGuard } from './auth-guard.service';
+import { NbPopoverTestComponent } from './popover-test/popover-test.component';
```
[the remark this answers is not in the document]
Comment (Tibing) [reply src/app/app.routes.ts]: fixed

File: src/framework/theme/components/popover/popover.component.scss (hunk 1 above)
Comment (Tibing) [reply src/framework/theme/components/popover/popover.component.scss]: fixed

File: src/framework/theme/components/popover/adjustment.helper.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
+import { NbPlacement, NbPositioningHelper } from './positioning.helper';
+
+export class NbPosition {
```
[the remark this answers is not in the document]
Comment (Tibing) [reply src/framework/theme/components/popover/adjustment.helper.ts]: fixed

File: src/framework/theme/components/popover/adjustment.helper.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+  position: {
+    top: number;
+    left: number;
+  };
+}
+
+/**
+ * Adjustment strategies.
+ * */
+export enum NbAdjustment {
```
[the remark this answers is not in the document]
Comment (Tibing) [reply src/framework/theme/components/popover/adjustment.helper.ts]: fixed

File: src/framework/theme/components/popover/adjustment.helper.ts [hunk 8]
```diff
@@ -0,0 +1,115 @@
+import { NbPlacement, NbPositioningHelper } from './positioning.helper';
+
+export class NbPosition {
+  placement: NbPlacement;
```
[the remark this answers is not in the document]
Comment (Tibing) [reply src/framework/theme/components/popover/adjustment.helper.ts]: fixed

File: src/framework/theme/components/popover/popover.directive.ts (hunk 2 above)
Comment (Tibing) [reply src/framework/theme/components/popover/popover.directive.ts]: fixed

File: src/framework/theme/components/popover/popover.directive.ts (hunk 3 above)
Comment (Tibing) [reply src/framework/theme/components/popover/popover.directive.ts]: fixed

File: src/framework/theme/components/popover/trigger.helper.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+   * @param getContainer {Function} popover container getter.
+   *
+   * @return {NbPopoverTrigger} open and close events streams.
+   * */
+  [NbPopoverMode.CLICK](host: HTMLElement, getContainer: Function): NbPopoverTrigger {
+    return {
+      open: observableFromEvent(host, 'click'),
+      close: observableFromEvent<Event>(document, 'click')
+        .pipe(
+          filter(event => !host.contains(event.target as Node)),
```
Comment (KostyaDanovsky) [inline src/framework/theme/components/popover/trigger.helper.ts]: Use one filter instead of 3

**Comment by nnixaa** [comment]
@Tibing add the component here also https://akveo.github.io/nebular/#/docs/concepts/ui-kit

File: src/framework/theme/components/popover/trigger.helper.ts (hunk 9 above)
Comment (Tibing) [reply src/framework/theme/components/popover/trigger.helper.ts]: fixed

File: src/framework/theme/components/popover/popover.directive.ts [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+  }
+
+  /*
+   * Subscribe to the popover triggers created from the {@link NbPopoverDirective#mode}.
+   * see {@link NbTriggerHelper}
+   * */
+  private registerTriggers() {
+    const { open, close } = NbTriggerHelper.createTrigger(this.hostElement, () => this.containerRef, this.mode);
+
+    open.pipe(takeWhile(() => this.alive))
```
[the remark this answers is not in the document]
Comment (Tibing) [reply src/framework/theme/components/popover/popover.directive.ts]: fixed

File: src/framework/theme/components/popover/popover.directive.ts [hunk 11]
```diff
@@ -0,0 +1,308 @@
+/**
```
[the remark this answers is not in the document]
Comment (Tibing) [reply src/framework/theme/components/popover/popover.directive.ts]: added

File: src/framework/theme/components/popover/popover.component.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
+/**
```
[the remark this answers is not in the document]
Comment (Tibing) [reply src/framework/theme/components/popover/popover.component.ts]: OnPush applied

[the remark this answers is not in the document]
Comment (nnixaa) [reply src/framework/theme/components/popover/popover.component.ts]: @Tibing have you checked it? As it looks to me that it won't render a component inside of it unless you check the input params

File: src/framework/theme/components/popover/_popover.component.theme.scss [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    border: 2px solid nb-theme(popover-border);
+    background: nb-theme(popover-bg);
+    box-shadow: nb-theme(popover-shadow);
+
+    .primitive-popover {
+      color: nb-theme(popover-fg);
+    }
+
+    .arrow {
+      border-bottom: 11px solid nb-theme(popover-border) !important;
```
[the remark this answers is not in the document]
Comment (Tibing) [reply src/framework/theme/components/popover/_popover.component.theme.scss]: fixed

**Pull request merged**